buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -278,39 +276,132 @@ open class DialectGenerateTask @Inject constructor(
+     * assumed that parsing begins at the start of a curly block. Maintains the
+     * state of which structure the parser is currently in to ensure that
+     * encountered curly braces are valid.
      */
-    private fun determineTokenValidity(
-        insideString: Boolean,
-        insideCharacter: Boolean,
-        insideSingleLineComment: Boolean,
-        insideMultiLineComment: Boolean
-    ) {
-        validBrace = !insideString &&
-                !insideSingleLineComment &&
-                !insideMultiLineComment &&
-                !insideCharacter
+    class CurlyParser() {
+
+        /**
+         * All of the possible important structures that a token can be inside of.
+         * This is tracked to ensure that when a curly brace is encountered,
+         * whether it is part of the syntax or not can be determined.
+         */
+        enum class InsideState {
 
-        validDoubleQuote = !insideCharacter &&
-                !insideSingleLineComment &&
-                !insideMultiLineComment
+            /**
+             * Not inside of anything.
+             */
+            NONE,
 
-        validSingleQuote = !insideString &&
-                !insideSingleLineComment &&
-                !insideMultiLineComment
+            /**
+             * Inside of a string declaration (block of double quotes).
+             */
+            STRING,
+
+            /**
+             * Inside of a character declaration (block of single quotes)
+             */
+            CHARACTER,
+
+            /**
+             * Inside of a single line comment.
+             */
+            SINGLE_COMMENT,
+
+            /**
+             * Inside of a multi line comment.
+             */
+            MULTI_COMMENT
+        }
 
-        validSingleComment = !insideString &&
-                !insideCharacter &&
-                !insideMultiLineComment
+        // Default state is that the parser is not within any structure.
+        private var insideState = InsideState.NONE
 
-        validMultiComment = !insideString &&
-                !insideCharacter &&
-                !insideSingleLineComment
+        // Keeps track of the number of open curly braces that have been
+        // legally encountered (those that are not within any structure).
+        private var curlyCounter = 0
+
+        /**
+         * Parses the given token and updates the state. It is assumed that the
+         * tokens are provided as a sequence:
+         *
+         * for any token1, token2
+         * if parseToken(token1) is called and parseToken(token2) is called right after,
+         * then token2 comes directly after token1 in the stream of tokens.
+         *
+         * @param token The token to parse
+         * @return Whether or not the curly block has been fully parsed
+         */
+        fun parseToken(token: String): Boolean {
+            insideState = getUpdatedState(token)
```

Comment:
nit: Suggest updating the state directly in getUpdatedState() since it has access to the insideState member variable and is already operating on it there. Also, consider renaming to just "updateState()".